hibernate-core/src/main/java/org/hibernate/sql/results/internal/domain/CircularBiDirectionalFetchImpl.java

```diff
@@ -182,11 +181,14 @@ public Object assemble(RowProcessingState rowProcessingState, JdbcValuesSourcePr
 			final Object initializedInstance = initializer.getInitializedInstance();
 			final LazyInitializer lazyInitializer = HibernateProxy.extractLazyInitializer( initializedInstance );
 			if ( lazyInitializer != null ) {
-				if ( initializedInstance.getClass().isAssignableFrom( javaType.getJavaTypeClass() ) ) {
+				final Class<?> concreteProxyClass = initializer.getConcreteDescriptor().getConcreteProxyClass();
+				if ( concreteProxyClass.isInstance( initializedInstance ) ) {
 					return initializedInstance;
```

Comment:
Not sure if that can happen, but I believe we should add the following logic:

```suggestion
					if ( timing == FetchTiming.IMMEDIATE ) {
						initializer.initializeInstance( rowProcessingState );
					}
					return initializedInstance;
```

Alternatively, we should at least add an assert:

```suggestion
					assert timing != FetchTiming.IMMEDIATE;
					return initializedInstance;
```

Reply:
if the proxy has not already been initialized it will done later, we only need to call `initializer.initializeInstance( rowProcessingState );` when we need to return the proxy initialized instance and not the proxy.